bot.py

```diff
@@ -133,6 +166,144 @@ async def info(self, ctx: commands.Context[Any]) -> None:
+
+        results.sort(key=(lambda r: r.average_time or Picoseconds(5e12)))
+        results = results[:10]
+        results.sort(key=(lambda r: r.id))
+
+        attachments = []
+        for res in results:
+            name = f"Submission_{res.id}.rs"
+            desc = f"Submission {res.id} from user {user}"
+            file_handle = StringIO(res.code)
+
+            # Not sure whether the type is specified incorrectly or this is an ugly hack.
+            # Either way, this is what works to get Discord to not open a file.
+            f = discord.File(file_handle, filename=name, description=desc)  # type: ignore[arg-type]
+            attachments.append(f)
+
+        await interaction.edit_original_response(
+            content=f"Ten fastest submissions for user {user} on day {day}, part {part}:",
+            attachments=attachments,
+        )
+
+    @app_commands.command()
+    @app_commands.default_permissions(manage_messages=True)
+    @only_from_guilds(*settings.discord.management_servers)
+    async def leaderboard_code(
+        self,
+        interaction: discord.Interaction,  # type: ignore[type-arg]
+        day: Annotated[AdventDay, app_commands.Range[int, 1, 25]],
+        part: Annotated[AdventPart, Literal[1, 2]],
+        place: Optional[app_commands.Range[int, 1, 10]] = None,
+    ) -> None:
+        await interaction.response.send_message(content="Loading...")
+
+        logger.info(
+            "User %s requested code for leaderboard entry (day %s, part %s, place %s)",
+            interaction.user,
+            day,
+            part,
+            place,
+        )
+
+        with Database() as db:
+            submisssions = db.get_lb_submissions(lib.year(), day, part)
+
+        if place is not None:
+            submisssions = [submisssions[place - 1]]
+
+        attachments = []
+        for res in submisssions:
+            user = self.bot.get_user(res.user_id) or await self.bot.fetch_user(res.user_id)
+            name = f"Submission_{user}_{res.id}.rs"
+            desc = f"Submission {res.id} from user {user}"
+            file_handle = StringIO(res.code)
+
+            # Not sure whether the type is specified incorrectly or this is an ugly hack.
+            # Either way, this is what works to get Discord to not open a file.
+            f = discord.File(file_handle, filename=name, description=desc)  # type: ignore[arg-type]
+            attachments.append(f)
+
+        await interaction.edit_original_response(
+            content=f"Leaderboard submissions on day {day}, part {part}:",
+            attachments=attachments,
+        )
+
+    @app_commands.command()
+    @app_commands.default_permissions(manage_messages=True)
+    @only_from_guilds(*settings.discord.management_servers)
+    async def invalidate_code(
+        self,
+        interaction: discord.Interaction,  # type: ignore[type-arg]
+        submission_id: Annotated[SubmissionId, int],
+    ) -> None:
```

Comment:
We might want to require another data point or two as a safety check to minimize typos.